### ISSUE TITLE
Instance#toJSON doesn't apply included models customized toJSON

### ISSUE THREAD

**Issue by DanMMX** [sequelize/sequelize#3891]
Using eager loading, I have a custom `toJSON` function to prevent important data from `Users` to go outside backend.

```
toJSON: function () {
    var values = this.get();
    delete values.someValue;
    delete values.anothervalue;
    return values;
}
```

When doing `Product.findAll({include: [User]})` that custom toJSON from User doesn't get called.

I did a custom `toJSON` into `Product` to prevent included `Users` to go with those values but the problem is not only `Product` needs eager load with `User` but also another 5 models.

Is there any way to use custom `toJSON` on `User` when including it on eager load?

Here is the custom `Product#toJSON` function

```
toJSON: function () {
    var values = this.get();
    if (values.User) {
        delete values.User.someValue;
        delete values.User.anotherValue;
    }
    return values;
}
```

**Comment by mickhansen** [comment]
How exactly are you defining your custom toJSON method? I.e. how are you attaching it to the instance.

**Comment by mickhansen** [comment]
But generally `JSON.stringify` will call `toJSON`, Sequelizes built-in `toJSON` will call `get({plain: true})` which will recursively call `get({plain: true})` on all included.

For blacklisting attributes you might be better off using something like: https://github.com/mickhansen/ssacl-attribute-roles

**Comment by DanMMX** [comment]
The way I'm doing `toJSON` definition is the following

```
var User = sequelize.define('User', {
	// attrs
}, {
	instanceMethods: {
		toJSON: function () {
			var values = this.get();
			delete values.someValue;
			delete values.anotherValue;
			return values;
		}
	},
	hooks: {
		// some hooks
	}
});
```

So then there's not a way to make it without `ssacl`? I mean, I don't want those 2 values in ANY case, they're used only for auth and shouldn't be available for anyone (well, only for that auth part…)

**Comment by mickhansen** [comment]
Sure, you can still overwrite `toJSON`, but just remember it isn't necessarily called if the child of another instance. In that case you'll have to overwrite the parent toJSON aswell so it doesn't call `get` like Sequelize does by default.

**Comment by arivia** [comment]
I wanted to hide certain attributes only when the data was sent back to the client while having access to the attributes on the instance of the model.

The code below worked for me.

**user.model.js**
```
const Sequelize = require('sequelize')
const Model = Sequelize.Model

const PROTECTED_ATTRIBUTES = ['password', 'token']

class User extends Model {
  toJSON () {
    // hide protected fields
    let attributes = Object.assign({}, this.get())
    for (let a of PROTECTED_ATTRIBUTES) {
      delete attributes[a]
    }
    return attributes
  }
}

User.init({
  email: Sequelize.STRING,
  firstName: Sequelize.STRING,
  lastName: Sequelize.STRING,
  password: Sequelize.STRING,
  token: Sequelize.STRING
},
{
  sequelize,
  modelName: 'user',
})

module.exports = User
```

**Comment by arivia** [comment]
We wanted access to the instance attributes at runtime but remove them before sending the data to the client. See this Gist on Github for a working solution
https://gist.github.com/arivia/2afbf227595fb5db24149e69fa8736e7

**Comment by lukeberry99** [comment]
@arivia this will not work when accessing the model through a relationship, any idea on how to solve that?

**Comment by ajmalahd** [comment]
@lukeberry99 I was able to solve this by adding a getter method to the field that returns undefined, like below:
```
firstName: {
      type: DataTypes.STRING,
      get() {
        return undefined;
      }
    }
```

See my answer on SO for more details - https://stackoverflow.com/a/59337434/1346528

**Comment by papb** [comment]
I think it might be a good idea to change the current Sequelize behavior to recursively respect custom `toJSON`s that could be defined around.

Sorry to bother you @mickhansen, do you remember why not to support this? Is there any downside?

In other words, why not change the current behavior, as shown by you:

> But generally `JSON.stringify` will call `toJSON`, Sequelizes built-in `toJSON` will call `get({plain: true})` which will recursively call `get({plain: true})` on all included.

To recursively call `toJSON` instead of calling `get({plain: true})`?

Recursively calling `toJSON` would, in turn, automatically call `get({plain: true})` anyway, unless a custom `toJSON` was defined there...

**Comment by lukas1994** [comment]
Are you working on this? Right now I'm doing sth like:
```
public toJSON() {
    const dcs: any[] = [];
    if (this.data_connections !== undefined) {
      this.data_connections.forEach(dc => {
        const dss: any[] = [];
        if (dc.data_sources !== undefined) {
          dc.data_sources.forEach(ds => dss.push(ds.toJSON()));
        }
        dc.data_sources = undefined;
        const x = dc.toJSON();
        x.data_sources = dss;
        dcs.push(x);
      });
    }
    this.data_connections = undefined;
    const out = this.get()
    out.data_connections = dcs;
    return out;
  }
```
which isn't ideal. Are there any better way of calling toJSON() recursively on all included models?

**Comment by lukas1994** [comment]
@papb could you please provide a more general solution for calling toJSON recursively on all included objects? that would make my life much easier

thanks!

**Comment by johanleroux** [comment]
A simple solution to this problem that follows the structure of Sequelize, is to add either an array called `hidden` to the `modelOptions` section of the model definition or a boolean `hidden` field on the attribute definition on a model.

Then by updating the internal `lib/model.js:get()` function we can hide by default all fields which are defined as hidden, even when including the relationships of a model through `get({ plain: true})`. This will also work when calling `toJSON` as it calls `get({ plain: true }) internally`.

Model definition
```js
module.exports = (sequelize, DataTypes) => {
  const user = sequelize.define('user', {
    email: DataTypes.STRING,
    password: DataTypes.STRING,
  }, { hidden: ['password'] })
  return user
}
```

Return model and hide `password` by default
```js
const user = User.findOne()
return user.get()
```

Return model and show `password`
```js
const user = User.findOne()
return user.get({ hidden: false })
```

Return `password` directly
```js
const user = User.findOne()
return user.get('password')
```

I have played around with this solution locally and got a prototype working, think this will be a great non-breaking feature that can be included in the core

**Comment by lukas1994** [comment]
In my use-case I don't just want to hide fields, I just want to use a custom toJSON() function that is used for included instances as well.

**Comment by Hakhenaton** [comment]
Here is what i did to solve this issue:

We use a base model class. All of our models extend from this abstract class.

Inside, we override both `toJSON` and `get` methods.

The main idea is to check in the overriden `get` method if we are in a `toJSON` serialization context.

Notice the extra parameter that we add in the get `options` object.

Also we must delete the unwanted fields only when the `get` method is called for retrieving all fields, and not a unique field.

So we simply check the arguments, if a string property is provided, then we just return the original return value.

Otherwise, we check if we have an options objet, and if it contains our special "toJSON" option. If it is the case, then we can safely delete the `removeFromJson` fields.

```typescript
abstract class BaseModel extends Model {

    public abstract readonly removeFromJson: (keyof this)[]
    
    public readonly id!: string
    public readonly createdAt!: Date
    public readonly updatedAt!: Date

    public get(...args: any[]): unknown {
        
    	const o = super.get(...args)

    	if (args.length > 1 || typeof args[0] === 'string')
    		return o

    	const options = args[0]

    	if (options === undefined)
    		return o

    	if (options.toJSON === true)
    		for (const field of this.removeFromJson)
    			delete o[field]

    	return o
    }

    public toJSON(): Partial<this> {

    	const json = _.cloneDeep(this.get({ plain: true, toJSON: true })) as Partial<this>

    	return json
    }
}
```

We need to declare our non-serializable fields in our sub classes, like this:

```typescript

import BaseModel from './BaseModel'

class MyModel extends BaseModel {

    public readonly removeFromJson: (keyof this)[] = [ 'secret' ]

    public readonly foo!: string
    public readonly bar! string
    public readonly secret!: string
}
```

Now the toJSON method will delete every field included in `removeFromJson`, and recursively in nested instances of other models (for as long as they extend the BaseModel class as well of course).

What do you think ?

**Comment by imjordanxd** [comment]
Any update on this?

**Comment by nermolov** [comment]
I agree that there should be built-in support for calling `toJSON` recursively through associations so that custom implementations can get used. In our case we're using a custom `toJSON` to support date-only fields as JS Date in API handlers but get converted to a date-only string for JSON, but that's broken with associations now.
